sqldelight-runtime/src/main/java/com/squareup/sqldelight/Query.kt

```diff
@@ -0,0 +1,93 @@
+package com.squareup.sqldelight
+
+import android.arch.persistence.db.SupportSQLiteOpenHelper
+import android.arch.persistence.db.SupportSQLiteQuery
+import android.database.Cursor
+
+/**
+ * A listenable, tyoed query generated by SQLDelight.
+ *
+ * @param RowType the type that this query can map it's result set to.
+ */
+abstract class Query<out RowType>(
+    private val openHelper: SupportSQLiteOpenHelper,
+    private val queryQueue: MutableList<Query<*>>,
+    private val mapper: (Cursor) -> RowType
+): SupportSQLiteQuery {
+  private val listeners = mutableSetOf<Listener>()
+
+  /**
+   * Notify listeners that their current result set is staled.
+   *
+   * Called internally by SQLDelight when it detects a possible staling of the result set. Emits
+   * some false positives but never misses a true positive.
+   */
+  fun notifyResultSetChanged() {
+    listeners.forEach(Listener::queryResultsChanged)
```

Comment:
This loop can execute concurrently with an add/remove below. It either needs to be guarded with a lock or use a CopyOnWriteArrayList (in which case you can't do add/isEmpty or remove/isEmpty in two separate operations).